File: 74.search-a-2-d-matrix.js
/*
 * @lc app=leetcode id=74 lang=javascript
 *
 * [74] Search a 2D Matrix
 */

// @lc code=start
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function (matrix, target) {
  const m = matrix.length;
  const n = matrix[0].length;
  let i = 0,
    j = m - 1;
  let res = m;
  // highbound:  first that > target
  while (i <= j) {
    let m = Math.floor(i + (j - i) / 2);
    if (matrix[m][0] <= target) {
      i = m + 1;
    } else {
      j = m - 1;
      res = m;
    }
  }
  if (res - 1 < 0) return false;

  i = 0;
  j = n - 1;
  while (i <= j) {
    let m = Math.floor(i + (j - i) / 2);
    if (matrix[res - 1][m] < target) {
      i = m + 1;
    } else if (matrix[res - 1][m] > target) {
      j = m - 1;
    } else {
      return true;
    }
  }
  return false;
};
// @lc code=end
searchMatrix([[1]], 1);

File: solutions/74.search-a-2d-matrix.rb
# @param {Integer[][]} matrix
# @param {Integer} target
# @return {Boolean}
def search_matrix(matrix, target)
  def search_higher(list, target)
    i = 0
    j = list.length - 1
    while i <= j
      m = i + (j - i) / 2
      if list[m] <= target
        i = m + 1
      else
        j = m - 1
        found = m
      end
    end
    found
  end

  def search(list, target)
    i = 0
    j = list.length - 1
    while i <= j
      m = i + (j-i)/2
      i = m+1 if list[m] < target
      j = m-1 if list[m] > target
      return m if list[m] == target
    end
  end

  row_higher_index = search_higher( matrix.map { _1[0] }, target) || matrix.length
  return false if (row_higher_index - 1).negative?

  search(matrix[row_higher_index-1], target).nil? ? false : true
end
